swaybar/ipc.c

```diff
@@ -322,8 +335,11 @@ void ipc_bar_init(struct bar *bar, const char *bar_id) {
 	}
 	free(res);
 	json_object_put(outputs);
-
-	const char *subscribe_json = "[ \"workspace\", \"mode\" ]";
+	// Will need to be redone for display mode toggling compatibility
+	// Will require setting bar->config->hidden_state to "show" as default
+	const char *subscribe_json = strcmp(bar->config->display_mode, "hide") == 0 ?
+		"[ \"workspace\", \"mode\", \"modifier\" ]"
+		: "[ \"workspace\", \"mode\" ]";
```

Comment:
>I don't think security mandates one approach over another, the client must be free to ask for whatever it likes, It's the server that enforces policy and the client should deal. I think (and so did @calvinkosmatka, independently) that having the client ask for everything it wants, but use only what the server gives corresponds perfectly to the security model and would simplify matters.

The server returns errors if you try to bind to events you don't have permission for. So either that needs to be handled with a retry mechanism that removes the unpermitted events, disables the relevant UI in the bar, and tries again. But I don't think that's important, can just ask for only what the user needs instead. We shouldn't be binding to events we aren't going to use anyway, if there's a bug that leads to that information being exposed from swaybar it reduces the impact of that vulnerability if it didn't bind to some events in the first place.

>This raises another point about client handling of security denials. In #1172 I saw that the server declined IPC, and swaybar carried on as if nothing happened, which I think is a problem.

Yes, the clients need to handle permissions errors more gracefully. I have done this for swaymsg but haven't gotten around to it for everyone else. Patches welcome

>That suggests that maybe the client should not subscribe to all events at once, but group them by their "severity". Some events are optional (modifier events), and some may be critical (abort if not available).

No, i3 compatability

>Wait, ok I see code in commands/bar/mode.c for handling bar display mode commands that sends an ipc barconfig event but the event never makes it to the socket because there's no security mapping there. I guess the question is whether to make bars subscribed to barconfig or bindings as a whole to do display mode switching.

I don't think the barconfig is maintained in the compositor beyond the initial allocation, but that's probably the best approach.